discovery/discovery.go

```diff
@@ -143,26 +143,72 @@ func doDiscover(app App, pre string, insecure bool) (*Endpoints, error) {
 		}
 	}
 
-	if len(de.ACIEndpoints) == 0 {
-		return nil, fmt.Errorf("found no ACI meta tags")
-	}
-
 	return de, nil
 }
 
-// DiscoverEndpoints will make HTTPS requests to find the ac-discovery meta
-// tags and optionally will use HTTP if insecure is set. Based on the app
-// passed the discovery templates will be filled out and returned in eps.
-func DiscoverEndpoints(app App, insecure bool) (eps *Endpoints, err error) {
+// DiscoverWalk will make HTTPS requests to find discovery meta tags and
+// optionally will use HTTP if insecure is set. Based on the response of the
+// discoverFn it will continue to recurse up the tree.
+func DiscoverWalk(app App, insecure bool, discoverFn DiscoverWalkFunc) (eps *Endpoints, err error) {
 	parts := strings.Split(string(app.Name), "/")
 	for i := range parts {
 		end := len(parts) - i
 		pre := strings.Join(parts[:end], "/")
-		eps, err = doDiscover(app, pre, insecure)
-		if err == nil {
-			break
+
+		eps, err = doDiscover(pre, app, insecure)
+		derr := discoverFn(pre, eps, err)
+		if derr != nil {
+			return
+		}
+	}
+
+	return
+}
+
+// DiscoverWalkFunc can stop a DiscoverWalk by returning non-nil error.
+type DiscoverWalkFunc func(prefix string, eps *Endpoints, err error) error
+
+// FailedAttempt represents a failed discovery attempt. This is for debugging
+// and user feedback.
+type FailedAttempt struct {
+	Prefix string
+	Error  error
+}
+
+// DiscoverEndpoints will make HTTPS requests to find the ac-discovery meta
+// tags and optionally will use HTTP if insecure is set. It will not give up
+// until it has exhausted the path or found an image discovery.
+func DiscoverEndpoints(app App, insecure bool) (eps *Endpoints, attempts []FailedAttempt, err error) {
+	dfn := func(pre string, eps *Endpoints, err error) error {
+		if err != nil {
+			attempts = append(attempts, FailedAttempt{pre, err})
+			return nil
+		}
+		if len(eps.ACIEndpoints) == 0 {
+			return fmt.Errorf("found no ACI meta tags")
```

Comment:
Yea, this is how the old code worked too. So, it would assume that the first 200 website it got should have meta info.